cloudinit/distros/__init__.py

```diff
@@ -961,6 +961,25 @@ def get_tmp_exec_path(self) -> str:
             return tmp_dir
         return os.path.join(self.usr_lib_exec, "cloud-init", "clouddir")
 
+    def do_as(self, command: list, user: str, cwd: str = "", **kwargs):
```

Comment:
Nice, I've never had a use for `preexec_fn` before but that looks better than what I did here.

How would you feel about allowing `subp()` to pass `**kwargs` directly to `Popen`?

Reply:
In this particular case, I think it'd be more convenient to have a `user=` argument and translate that to the proper UID. As of 3.9, POpen has that...but...3.9.

I'm not against the idea in general, but I'm a little wary of it getting confusing if something gets passed via a defined keyword arg plus subp processing vs the kwargs dict.

Reply:
I was working on an implementation of this, but unfortunately `preexec_fn` isn't safe since it [could lead to deadlock](https://discuss.python.org/t/why-preexec-fn-in-subprocess-popen-may-lead-to-deadlock/16908).

Looking around at `os` and `sys`, I don't see much for pure-language options that are also high level interfaces (fork/exec are all I see at the moment).

How would we feel about using fork/exec? If we would prefer that over the current approach I can try that.

#### Some other possibilities I've ruled out:

Looking at the `Popen` user/group implementation in Python 3.9+, I see that the python library implementation depends on an internal function `fork_exec` imported from `_posixsubprocess`. By the naming I don't think emulating the python library is the safest approach.

A context manager that sets the current pid/gid before running `Popen()` wouldn't work, because the context might not have permissions to `setuid()` back to `1`.

Reply:
Per out of band discussion, current proposal is to use what we have, and explore pure-language options further in a followup PR.